File: typescript/src/prices.ts
import express from "express";
import {mimicDbConnection} from "./db";

async function createApp() {
    const app = express()

    const connection = await mimicDbConnection();

    app.put('/prices', async (req, res) => {
        // This does nothing, as we don't write to DB.
        res.json()
    })
    app.get('/prices', async (req, res) => {
        // @ts-ignore
        const result = await connection.readPrices(req.query.type)
        if (req.query.age as any < 6) {
            res.json({cost: 0})
        } else {
            if (req.query.type !== 'night') {
                const holidays = await connection.readHolidays()

                let isHoliday;
                let reduction = 0
                for (let row of holidays) {
                    let holiday = row.holiday
                    if (req.query.date) {
                        let d = new Date(req.query.date as string)
                        if (d.getFullYear() === holiday.getFullYear()
                            && d.getMonth() === holiday.getMonth()
                            && d.getDate() === holiday.getDate()) {

                            isHoliday = true
                        }
                    }

                }

                if (!isHoliday && new Date(req.query.date as string).getDay() === 1) {
                    reduction = 35
                }

                // TODO apply reduction for others
                if (req.query.age as any < 15) {
                    res.json({cost: Math.ceil(result.cost * .7)})
                } else {
                    if (req.query.age === undefined) {
                        let cost = result.cost * (1 - reduction / 100)
                        res.json({cost: Math.ceil(cost)})
                    } else {
                        if (req.query.age as any > 64) {
                            let cost = result.cost * .75 * (1 - reduction / 100)
                            res.json({cost: Math.ceil(cost)})
                        } else {
                            let cost = result.cost * (1 - reduction / 100)
                            res.json({cost: Math.ceil(cost)})
                        }
                    }
                }
            } else {
                if (req.query.age as any >= 6) {
                    if (req.query.age as any > 64) {
                        res.json({cost: Math.ceil(result.cost * .4)})
                    } else {
                        res.json(result)
                    }
                } else {
                    res.json({cost: 0})
                }
            }
        }
    })
    return {app, connection}
}

export {createApp}

File: python/src/prices.py
import json
import math
import os
from datetime import datetime

from flask import Flask
from flask import request

from src.db import mimic_db_operations

app = Flask("lift-pass-pricing")

db_like = None

f = open(os.path.join(os.path.realpath(
    os.path.join(os.getcwd(), os.path.dirname(__file__))), "prices_db.json"), "r")

rawdata = json.load(f)
data = {
    "prices": rawdata.get("prices"),
    "holidays": [datetime.fromisoformat(x) for x in rawdata.get("holidays")]
}


@app.route("/prices", methods=['GET', 'PUT'])
def prices():
    res = {}
    global db_like
    if db_like is None:
        db_like = mimic_db_operations()
    if request.method == 'PUT':
        return {}
    elif request.method == 'GET':
        result = db_like.read_price( request.args['type'])

        if 'age' in request.args and request.args.get('age', type=int) < 6:
            res["cost"] = 0
        else:
            if "type" in request.args and request.args["type"] != "night":
                is_holiday = False
                reduction = 0
                list_of_holidays = db_like.list_holidays()
                for holiday in list_of_holidays:
                    if "date" in request.args:
                        d = datetime.fromisoformat(request.args["date"])
                        if d.year == holiday.year and d.month == holiday.month and holiday.day == d.day:
                            is_holiday = True
                if not is_holiday and "date" in request.args and datetime.fromisoformat(
                        request.args["date"]).weekday() == 0:
                    reduction = 35

                # TODO: apply reduction for others
                if 'age' in request.args and request.args.get('age', type=int) < 15:
                    res['cost'] = math.ceil(result["cost"] * .7)
                else:
                    if 'age' not in request.args:
                        cost = result['cost'] * (1 - reduction / 100)
                        res['cost'] = math.ceil(cost)
                    else:
                        if 'age' in request.args and request.args.get('age', type=int) > 64:
                            cost = result['cost'] * .75 * (1 - reduction / 100)
                            res['cost'] = math.ceil(cost)
                        elif 'age' in request.args:
                            cost = result['cost'] * (1 - reduction / 100)
                            res['cost'] = math.ceil(cost)
            else:
                if 'age' in request.args and request.args.get('age', type=int) >= 6:
                    if request.args.get('age', type=int) > 64:
                        res['cost'] = math.ceil(result['cost'] * .4)
                    else:
                        res.update(result)
                else:
                    res['cost'] = 0

    return res


if __name__ == "__main__":
    app.run(port=3005)
